src/Week5/Calc.hs

```diff
@@ -0,0 +1,308 @@
+
+ One way to resolve the ambiguity is by giving an explicit type signature, as in the
+ above example. Another way is by using such an expression as part of some larger
+ expression so that the context in which it is used determines the type.
+
+ For example, we may write a function reify as follows:
+    reify :: ExprT -> ExprT
+    reify = id
+
+ To the untrained eye it may look like reify does no actual work!
+ But its real purpose is to constrain the type of its argument to ExprT.
+ Now we can write things like
+    reify $ mul (add (lit 2) (lit 3)) (lit 4)
+ at the ghci prompt
+-}
+class Expr e where
+  lit :: Integer -> e
+  add :: e -> e -> e
+  mul :: e -> e -> e
+
+
+instance Expr ExprT where
+  lit = Lit
+  add = Add
+  mul = Mul
+
+
+{-
+ Exercise 4:
+
+ The marketing department has gotten wind of just how flexible
+ the calculator project is and has promised custom calculators to some
+ big clients. As you noticed after the initial roll-out, everyone loves the
+ interface, but everyone seems to have their own opinion on what the
+ semantics should be. Remember when we wrote ExprT and thought
+ that addition and multiplication of integers was pretty cut and dried?
+ Well, it turns out that some big clients want customized calculators
+ with behaviors that they have decided are right for them.
+
+ The point of our Expr type class is that we can now write down
+ arithmetic expressions once and have them interpreted in various
+ ways just by using them at various types.
+
+ Make instances of Expr for each of the following types:
+ • Integer — works like the original calculator
+
+ • Bool — every literal value less than or equal to 0 is interpreted as False,
+          and all positive Integers are interpreted as True; “addition” is logical or,
+          “multiplication” is logical and
+
+ • MinMax — “addition” is taken to be the max function, while “multiplication”
+            is the min function
+
+ • Mod7 — all values should be in the ranage 0 . . . 6, and all arithmetic is
+          done modulo 7; for example, 5 + 3 = 1.
+
+ The last two variants work with Integers internally, but in order
+ to provide different instances, we wrap those Integers in newtype
+ wrappers. These are used just like the data constructors we’ve seen before.
+
+   newtype MinMax = MinMax Integer deriving (Eq, Show)
+   newtype Mod7 = Mod7 Integer deriving (Eq, Show)
+-}
+instance Expr Integer where
+  lit = id
+  add = (+)
+  mul = (*)
+
+
+instance Expr Bool where
+  lit = (0 <)
+  add = (||)
+  mul = (&&)
+
+
+newtype MinMax = MinMax Integer deriving (Show, Eq)
+
+
+instance Expr MinMax where
+  lit n                     = MinMax n
+  add (MinMax x) (MinMax y) = MinMax $ max x y
+  mul (MinMax x) (MinMax y) = MinMax $ min x y
+
+
+newtype Mod7 = Mod7 Integer deriving (Show, Eq)
+
+
+instance Expr Mod7 where
+  lit n                 = Mod7 $ n `mod` 7
+  add (Mod7 x) (Mod7 y) = Mod7 $ (x + y) `mod` 7
+  mul (Mod7 x) (Mod7 y) = Mod7 $ (x * y) `mod` 7
```

Comment:
I think we can write this in terms of `lit` as well. Something like:
```haskell
add (Mod7 x) (Mod7 y) = lit $ x + y
mul (Mod7 x) (Mod7 y) = lit $ x * y
```

But I'm not sure if that's a preferred way to keep falling back to base cases in the instance definition itself.

Reply:
I think this is the better solution, since the `lit` function implicitly performs the `mod` 7 operation